src/main/java/com/hartwig/snpcheck/SnpCheck.java

```diff
@@ -64,49 +63,45 @@ public SnpCheck(final RunApi runs, final SampleApi samples, final Bucket snpchec
         this.labPendingBuffer = new LabPendingBuffer(this, Executors.newScheduledThreadPool(1), TimeUnit.HOURS, 1);
     }
 
-    public void handle(final PipelineStaged event) {
+    public void handle(final PipelineComplete event) {
         try {
-            if (!event.analysisContext().equals(Context.SHALLOW)) {
-                Run run = runs.get(event.runId().orElseThrow());
-                if (run.getStatus().equals(Status.VALIDATED)) {
-                    LOGGER.info("Passing through [{} {}] event for already-snpchecked run [{}]",
-                            event.analysisContext(), event.analysisType(), run.getId());
-                    PipelineValidated.builder().originalEvent(event).build().publish(validatedTopicPublisher, objectMapper);
-                }
-                if (run.getIni().equals(Ini.SOMATIC_INI.getValue()) || run.getIni().equals(Ini.SINGLESAMPLE_INI.getValue())) {
-                    LOGGER.info("Received a SnpCheck candidate [{}] for run [{}]", run.getSet().getName(), run.getId());
-                    if (waitForFinished(run)) {
-                        Iterable<Blob> valVcfs = Optional.ofNullable(snpcheckBucket.list(Storage.BlobListOption.prefix(SNPCHECK_VCFS)))
-                                .map(Page::iterateAll)
-                                .orElse(Collections.emptyList());
-                        Optional<Sample> maybeRefSample = onlyOne(samples, run.getSet(), SampleType.REF);
-                        Optional<Sample> maybeTumorSample = onlyOne(samples, run.getSet(), SampleType.TUMOR);
-                        if (maybeRefSample.isPresent()) {
-                            Sample refSample = maybeRefSample.get();
-                            Optional<Blob> maybeValVcf = findValidationVcf(valVcfs, refSample);
-                            if (maybeValVcf.isPresent()) {
-                                VcfComparison.Result result = doComparison(run, refSample, maybeValVcf.get());
-                                SnpCheckEvent.builder()
-                                        .publisher(turquoiseTopicPublisher)
-                                        .sample(maybeTumorSample.map(Sample::getName).orElse(refSample.getName()))
-                                        .result(result.name().toLowerCase())
+            Run run = runs.get(event.pipeline().runId());
+            if (run.getIni().equals(Ini.SOMATIC_INI.getValue()) || run.getIni().equals(Ini.SINGLESAMPLE_INI.getValue())) {
+                LOGGER.info("Received a SnpCheck candidate [{}] for run [{}]", run.getSet().getName(), run.getId());
+                if (waitForFinished(run)) {
```

Comment:
Will we encounter unfinished runs anymore now that all the events are together?

Reply:
Indeed we can!